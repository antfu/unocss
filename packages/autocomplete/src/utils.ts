export function searchUsageBoundary(line: string, index: number) {
  let start = index
  let end = index

  const regex = /[^\s>"'`;]/
  while (start && regex.test(line.charAt(start - 1))) --start
  while (end < line.length && regex.test(line.charAt(end))) ++end

  return {
    content: line.slice(start, end),
    start,
    end,
  }
}
