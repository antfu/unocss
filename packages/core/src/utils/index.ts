export * from './escape'
export * from './object'
export * from './basic'
export * from './helpers'
export * from './map'
export * from './set'
export * from './layer'
export * from './variantGroup'
export * from './warn'
export * from './handlers'
