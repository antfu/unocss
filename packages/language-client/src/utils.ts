import { MarkdownString } from 'vscode'

export function getMarkdown(code: string, lang?: string) {
  return new MarkdownString(getMarkdownCodeBlock(code, lang))
}

export function getMarkdownCodeBlock(code: string, lang = '') {
  return `\`\`\`${lang}\n${code}\n\`\`\``
}
