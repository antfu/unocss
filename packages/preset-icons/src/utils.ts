/* eslint-disable n/prefer-global/process */
export const isNode = typeof process < 'u' && typeof process.stdout < 'u' && !process.versions.deno
export const isVSCode = isNode && !!process.env.VSCODE_CWD
