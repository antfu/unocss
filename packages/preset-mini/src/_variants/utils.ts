export const PseudoClasses: Record<string, string> = Object.fromEntries([
  // pseudo elements part 1
  ['first-letter', '::first-letter'],
  ['first-line', '::first-line'],

  // location
  'any-link',
  'link',
  'visited',
  'target',
  ['open', '[open]'],

  // forms
  'default',
  'checked',
  'indeterminate',
  'placeholder-shown',
  'autofill',
  'optional',
  'required',
  'valid',
  'invalid',
  'user-valid',
  'user-invalid',
  'in-range',
  'out-of-range',
  'read-only',
  'read-write',

  // content
  'empty',

  // interactions
  'focus-within',
  'hover',
  'focus',
  'focus-visible',
  'active',
  'enabled',
  'disabled',
  'popover-open',

  // tree-structural
  'root',
  'empty',
  ['even-of-type', ':nth-of-type(even)'],
  ['even', ':nth-child(even)'],
  ['odd-of-type', ':nth-of-type(odd)'],
  ['odd', ':nth-child(odd)'],
  'first-of-type',
  ['first', ':first-child'],
  'last-of-type',
  ['last', ':last-child'],
  'only-child',
  'only-of-type',

  // pseudo elements part 2
  ['backdrop-element', '::backdrop'],
  ['placeholder', '::placeholder'],
  ['before', '::before'],
  ['after', '::after'],
  ['file', '::file-selector-button'],
].map(key => Array.isArray(key) ? key : [key, `:${key}`]))

export const PseudoClassesKeys = Object.keys(PseudoClasses)

export const PseudoClassesColon: Record<string, string> = Object.fromEntries([
  ['backdrop', '::backdrop'],
].map(key => Array.isArray(key) ? key : [key, `:${key}`]))

export const PseudoClassesColonKeys = Object.keys(PseudoClassesColon)

export const PseudoClassFunctions = [
  'not',
  'is',
  'where',
  'has',
]

export const PseudoClassesMulti: Record<string, string[]> = Object.fromEntries([
  ['selection', ['::selection', ' *::selection']],
  ['marker', ['::marker', ' *::marker']],
])

export const PseudoClassesStr = Object.entries(PseudoClasses)
  .filter(([, pseudo]) => !pseudo.startsWith('::'))
  .map(([key]) => key)
  .sort((a, b) => b.length - a.length)
  .join('|')
export const PseudoClassesColonStr = Object.entries(PseudoClassesColon)
  .filter(([, pseudo]) => !pseudo.startsWith('::'))
  .map(([key]) => key)
  .sort((a, b) => b.length - a.length)
  .join('|')
export const PseudoClassFunctionsStr = PseudoClassFunctions.join('|')
const PseudoClassesMultiKeys = Object.keys(PseudoClassesMulti)
export const PseudoClassesMultiStr = PseudoClassesMultiKeys.sort((a, b) => b.length - a.length).join('|')

export const PseudoClassesStrRe = new RegExp([...PseudoClassesKeys, ...PseudoClassesColonKeys, ...PseudoClassFunctions, ...PseudoClassesMultiKeys].join('|'))
export function isPseudoClass(input: string) {
  return PseudoClassesStrRe.test(input)
}
