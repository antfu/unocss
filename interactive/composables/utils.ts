import { uniq } from '@unocss/core'
import type { ResultItem, RuleItem } from '~/types'
import { mdnIndex } from '~/data/mdn-index'

export function extractColors(css: string) {
  return Array.from(css.matchAll(/\brgba?\((.+?)\)/g))
    .map((i) => {
      const [r, g, b, a] = i[1].split(',').map(i => parseInt(i.trim()))
      if (Number.isNaN(r))
        return ''
      if (!Number.isNaN(a))
        return `rgba(${r}, ${g}, ${b}, ${a})`
      return `rgb(${r}, ${g}, ${b})`
    })
    .filter(Boolean)
}

export function getFeatureUsage(css: string) {
  const props = uniq([...css.matchAll(/^\s+(\w[\w-]+)\:/mg)].map(i => i[1]))
  const functions = uniq([...css.matchAll(/\b(\w+)\(/mg)].map(i => `${i[1]}()`))
  return [...props, ...functions]
    .filter(i => mdnIndex.find(s => s.title === i))
}

export function getDocs(item: RuleItem) {
  return item.features?.map(i => mdnIndex.find(s => s.title === i)!) || []
}

export function getUtilsOfFeature(name: string) {
  return [...(featuresMap.get(name) || [])]
}

export function findAlias(item: RuleItem) {
  return [...matchedMap.values()].filter(i => i.body === item.body && i.class !== item.class)
}

export function goRandom() {
  const keys = Array.from(matchedMap.keys())
  const index = Math.round(keys.length * Math.random())
  useRouter().push({ query: { s: keys[index] } })
}

export function sampleArray<T>(arr: T[], count: number) {
  return Array.from({ length: count }, _ => arr[Math.round(Math.random() * (arr.length - 1))])
}

export function getItemId(item: ResultItem) {
  if (item.type !== 'rule')
    return `${item.type}:${item.title}`
  return item.class
}
