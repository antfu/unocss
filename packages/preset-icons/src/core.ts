import type { DynamicMatcher, Preset, Rule } from '@unocss/core'
import { warnOnce } from '@unocss/core'
import type {
  IconifyLoaderOptions,
  UniversalIconLoader,
} from '@iconify/utils/lib/loader/types'
import { encodeSvgForCss } from '@iconify/utils/lib/svg/encode-svg-for-css'

import { mergeIconProps, trimSVG } from '@iconify/utils'
import type { CSSSVGSprites, IconsOptions } from './types'
import { createAsyncSpriteIconsFactory, createUint8ArraySprite } from './create-sprite'

const COLLECTION_NAME_PARTS_MAX = 3

export { IconsOptions }

/* TODO BEGIN-CLEANUP: types from @iconify/utils: remove them once published */
interface GeneratedSpriteData {
  name: string
  asset: Uint8Array
}
/* TODO END-CLEANUP: types from @iconify/utils: remove them once published */

interface PresetIcons extends Preset {
  generateCSSSVGSprites: () => AsyncIterableIterator<GeneratedSpriteData>
  createCSSSVGSprite: (collection: string) => Promise<Uint8Array | undefined>
}

export function createPresetIcons(lookupIconLoader: (options: IconsOptions) => Promise<UniversalIconLoader>) {
  return function presetIcons(options: IconsOptions = {}): Preset {
    const {
      scale = 1,
      mode = 'auto',
      prefix = 'i-',
      warn = false,
      collections: customCollections,
      extraProperties = {},
      customizations = {},
      autoInstall = false,
      layer = 'icons',
      unit,
      sprites,
    } = options

    const loaderOptions: IconifyLoaderOptions = {
      addXmlNs: true,
      scale,
      customCollections,
      autoInstall,
      // avoid warn from @iconify/loader: we'll warn below if not found
      warn: undefined,
      customizations: {
        ...customizations,
        additionalProps: { ...extraProperties },
        trimCustomSvg: true,
        async iconCustomizer(collection, icon, props) {
          await customizations.iconCustomizer?.(collection, icon, props)
          if (unit) {
            if (!props.width)
              props.width = `${scale}${unit}`
            if (!props.height)
              props.height = `${scale}${unit}`
          }
        },
      },
    }

    const rules: Rule[] = [[
      /^([a-z0-9:_-]+)(?:\?(mask|bg|auto))?$/,
      createDynamicMatcher(warn, mode, loaderOptions, iconLoaderResolver),
      { layer, prefix },
    ]]

    if (sprites) {
      rules.push([
        /^([a-z0-9:_-]+)(?:\?(mask|bg|auto))?$/,
        createDynamicMatcher(warn, sprites.mode ?? mode, loaderOptions, iconLoaderResolver, sprites),
        { layer, prefix: sprites.prefix ?? 'sprite-' },
      ])
    }

    let iconLoader: UniversalIconLoader

    async function iconLoaderResolver() {
      iconLoader = iconLoader || await lookupIconLoader(options)
      return iconLoader
    }

    return <PresetIcons>{
      name: '@unocss/preset-icons',
      enforce: 'pre',
      options,
      layers: { icons: -30 },
      rules,
      generateCSSSVGSprites: createGenerateCSSSVGSprites(sprites),
      createCSSSVGSprite: createCSSSVGSpriteLoader(sprites),
    }
  }
}

export function combineLoaders(loaders: UniversalIconLoader[]) {
  return (async (...args) => {
    for (const loader of loaders) {
      if (!loader)
        continue
      const result = await loader(...args)
      if (result)
        return result
    }
  }) as UniversalIconLoader
}

function createDynamicMatcher(
  warn: boolean,
  mode: string,
  loaderOptions: IconifyLoaderOptions,
  iconLoaderResolver: () => Promise<UniversalIconLoader>,
  sprites?: CSSSVGSprites,
) {
  return <DynamicMatcher>(async ([full, body, _mode = mode]) => {
    let collection = ''
    let name = ''
    let svg: string | undefined

    const iconLoader = await iconLoaderResolver()

    const usedProps = {}
    if (body.includes(':')) {
      [collection, name] = body.split(':')
      svg = sprites
        ? await loadSvgFromSprite(collection, name, sprites, { ...loaderOptions, usedProps })
        : await iconLoader(collection, name, { ...loaderOptions, usedProps })
    }
    else {
      const parts = body.split(/-/g)
      for (let i = COLLECTION_NAME_PARTS_MAX; i >= 1; i--) {
        collection = parts.slice(0, i).join('-')
        name = parts.slice(i).join('-')
        svg = sprites
          ? await loadSvgFromSprite(collection, name, sprites, { ...loaderOptions, usedProps })
          : await iconLoader(collection, name, { ...loaderOptions, usedProps })
        if (svg)
          break
      }
    }

    if (!svg) {
      if (warn)
        warnOnce(`failed to load icon "${full}"`)
      return
    }

    let url: string
    if (sprites)
      url = `url("unocss-${collection}-sprite.svg#${name}")`
    else
      url = `url("data:image/svg+xml;utf8,${encodeSvgForCss(svg)}")`

    if (_mode === 'auto')
      _mode = svg.includes('currentColor') ? 'mask' : 'bg'

    if (_mode === 'mask') {
      // Thanks to https://codepen.io/noahblon/post/coloring-svgs-in-css-background-images
      return {
        '--un-icon': url,
        '-webkit-mask': 'var(--un-icon) no-repeat',
        'mask': 'var(--un-icon) no-repeat',
        '-webkit-mask-size': '100% 100%',
        'mask-size': '100% 100%',
        'background-color': 'currentColor',
        // for Safari https://github.com/elk-zone/elk/pull/264
        'color': 'inherit',
        ...usedProps,
      }
    }
    else {
      return {
        'background': `${url} no-repeat`,
        'background-size': '100% 100%',
        'background-color': 'transparent',
        ...usedProps,
      }
    }
  })
}

/* TODO BEGIN-CLEANUP: types from @iconify/utils: remove them once published */
function createCSSSVGSpriteLoader(sprites?: CSSSVGSprites) {
  return async (collection: string) => {
    const collections = sprites?.sprites[collection]
    if (!collections)
      return undefined

    return await createUint8ArraySprite(
      collection,
      createAsyncSpriteIconsFactory(collections),
      sprites?.warn ?? false,
    )
  }
}

function createGenerateCSSSVGSprites(sprites?: CSSSVGSprites) {
  return async function* () {
    if (sprites) {
      const warn = sprites.warn ?? false
      for (const [collection, collections] of Object.entries(sprites.sprites)) {
        yield <GeneratedSpriteData>{
          name: collection,
          asset: await createUint8ArraySprite(
            collection,
            createAsyncSpriteIconsFactory(collections),
            warn,
          ),
        }
      }
    }
  }
}

async function customizeSpriteIcon(
  collection: string,
  icon: string,
  options: IconifyLoaderOptions,
  svg?: string,
) {
  if (!svg)
    return svg

  const cleanupIdx = svg.indexOf('<svg')
  if (cleanupIdx > 0)
    svg = svg.slice(cleanupIdx)

  const { transform } = options?.customizations ?? {}
  svg
      = typeof transform === 'function'
      ? await transform(svg, collection, icon)
      : svg

  if (!svg.startsWith('<svg')) {
    console.warn(
        `Custom icon "${icon}" in "${collection}" is not a valid SVG`,
    )
    return svg
  }

  return await mergeIconProps(
    options?.customizations?.trimCustomSvg === true
      ? trimSVG(svg)
      : svg,
    collection,
    icon,
    options,
    undefined,
  )
}

async function loadSvgFromSprite(
  collectionName: string,
  icon: string,
  sprites: CSSSVGSprites,
  options: IconifyLoaderOptions,
) {
  const collections = sprites.sprites[collectionName]
  if (!collections)
    return

  const collectionsArray = Array.isArray(collections)
    ? collections
    : [collections]

  for (const collection of collectionsArray) {
    if (Array.isArray(collection)) {
      for (const spriteIcon of collection) {
        if (spriteIcon.name === icon)
          return await customizeSpriteIcon(collectionName, icon, options, spriteIcon.svg)
      }
    }
    else if ('svg' in collection) {
      if (collection.name === icon)
        return await customizeSpriteIcon(collectionName, icon, options, collection.svg)
    }
    else if (typeof collection === 'function') {
      for await (const spriteIcon of collection()) {
        if (spriteIcon.name === icon)
          return await customizeSpriteIcon(collectionName, icon, options, spriteIcon.svg)
      }
    }
    else {
      for await (const spriteIcon of collection[Symbol.asyncIterator]()) {
        if (spriteIcon.name === icon)
          return await customizeSpriteIcon(collectionName, icon, options, spriteIcon.svg)
      }
    }
  }
}
/* TODO END-CLEANUP: types from @iconify/utils: remove them once published */
