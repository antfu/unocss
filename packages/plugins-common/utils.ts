import { escapeRegExp as e, isAttributifySelector } from '@unocss/core'

export function getPath(id: string) {
  return id.replace(/\?.*$/, '')
}

// https://github.com/dsblv/string-replace-async/blob/main/index.js
export function replaceAsync(string: string, searchValue: RegExp, replacer: (...args: string[]) => Promise<string>) {
  try {
    if (typeof replacer === 'function') {
      const values: Promise<string>[] = []
      String.prototype.replace.call(string, searchValue, (...args) => {
        values.push(replacer(...args))
        return ''
      })
      return Promise.all(values).then((resolvedValues) => {
        return String.prototype.replace.call(string, searchValue, () => {
          return resolvedValues.shift() || ''
        })
      })
    }
    else {
      return Promise.resolve(
        String.prototype.replace.call(string, searchValue, replacer),
      )
    }
  }
  catch (error) {
    return Promise.reject(error)
  }
}

export function getMatchedPositions(code: string, matched: string[]) {
  const result: [number, number, string][] = []
  const attributify: RegExpMatchArray[] = []
  const plain = new Set<string>()

  Array.from(matched)
    .forEach((v) => {
      const match = isAttributifySelector(v)
      if (!match)
        plain.add(v)
      else if (!match[2])
        plain.add(match[1])
      else
        attributify.push(match)
    })

  // hightlight for plain classes
  let start = 0
  code.split(/[\s"';<>]/g).forEach((i) => {
    const end = start + i.length
    if (plain.has(i))
      result.push([start, end, i])
    start = end + 1
  })

  // attributify values
  attributify.forEach(([, name, value]) => {
    const regex = new RegExp(`(${e(name)}=)(['"])[^\\2]*?${e(value)}[^\\2]*?\\2`, 'g')
    Array.from(code.matchAll(regex))
      .forEach((match) => {
        const escaped = match[1]
        const body = match[0].slice(escaped.length)
        const bodyIndex = body.indexOf(value)
        if (bodyIndex < 0)
          return
        const start = match.index! + escaped.length + bodyIndex
        const end = start + value.length
        result.push([start, end, `[${name}="${value}"]`])
      })
  })

  return result
}
