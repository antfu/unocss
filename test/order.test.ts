import { createGenerator } from '@unocss/core'
import { describe, expect, test } from 'vitest'
import { variantMatcher } from '@unocss/preset-mini/utils'

describe('order', () => {
  test('static', async() => {
    const uno = createGenerator({
      rules: [
        ['foo', { name: 'bar1' }],
        ['foo', { name: 'bar2' }],
      ],
      presets: [],
    })
    const { css } = await uno.generate('foo')
    expect(css).toContain('bar2')
  })

  test('dynamic', async() => {
    const uno = createGenerator({
      rules: [
        [/^foo$/, () => ({ name: 'bar1' })],
        [/^foo$/, () => ({ name: 'bar2' })],
      ],
      presets: [],
    })
    const { css } = await uno.generate('foo')
    expect(css).toContain('bar2')
  })

  test('variant ordering', async() => {
    const uno = createGenerator({
      rules: [
        [/^foo-.$/, ([m]) => ({ name: m })],
      ],
      presets: [],
      variants: [
        variantMatcher('light', input => `.light $$ ${input}`),
        variantMatcher('group', input => `.group ${input}`),
        (v) => {
          const match = variantMatcher('dark', input => `.dark $$ ${input}`)(v)
          if (match) {
            return {
              ...match,
              order: 1,
            }
          }
        },
      ],
    })
    const code = [
      'light:group:foo-1',
      'group:light:foo-2',
      'dark:group:foo-3',
      'group:dark:foo-4',
    ].join(' ')
    const { css } = await uno.generate(code)
    const { css: css2 } = await uno.generate(code)
    expect(css).toMatchSnapshot()
    expect(css).toEqual(css2)
  })

  test('variant sorting', async() => {
    const uno = createGenerator({
      rules: [
        [/^foo-.$/, ([m]) => ({ foo: m })],
      ],
      presets: [],
      variants: [
        (input: string) => {
          const m = input.match(/^(\w+):/)
          if (m) {
            return {
              matcher: input.slice(m[0].length),
              selector: s => `${m[1]} ${s}`,
              sort: {
                pre: -1,
                post: 1,
              }[m[1]],
            }
          }
        },
      ],
    })
    const code = [
      'any:foo-1',
      'post:foo-1',
      'pre:foo-1',
      'zzz:foo-1',
    ].join(' ')
    const { css } = await uno.generate(code)
    const { css: css2 } = await uno.generate(code)
    expect(css).toMatchSnapshot()
    expect(css).toEqual(css2)
  })
})
