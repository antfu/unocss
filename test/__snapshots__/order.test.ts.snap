// Vitest Snapshot v1

exports[`order > movePseudoElementsEnd 1`] = `".marker\\\\:file\\\\:hover\\\\:selection\\\\:mb-4:hover::marker::file-selector-button::selection"`;

exports[`order > multiple variant sorting 1`] = `
"/* layer: default */
.dark .group:hover:focus-within .dark\\\\:group-hover\\\\:group-focus-within\\\\:bg-blue-600{--un-bg-opacity:1;background-color:rgba(37,99,235,var(--un-bg-opacity));}
.group:hover:focus-within .dark .group-hover\\\\:group-focus-within\\\\:dark\\\\:bg-red-600{--un-bg-opacity:1;background-color:rgba(220,38,38,var(--un-bg-opacity));}"
`;

exports[`order > variant ordering 1`] = `
"/* layer: default */
.dark .group .dark\\\\:group\\\\:foo-3{name:foo-3;}
.dark .group .group\\\\:dark\\\\:foo-4{name:foo-4;}
.group .light .group\\\\:light\\\\:foo-2{name:foo-2;}
.light .group .light\\\\:group\\\\:foo-1{name:foo-1;}"
`;

exports[`order > variant sorting 1`] = `
"/* layer: default */
pre .pre\\\\:foo-1,
any .any\\\\:foo-1,
zzz .zzz\\\\:foo-1,
post .post\\\\:foo-1{foo:foo-1;}"
`;
